src/java.base/share/classes/sun/util/locale/BaseLocale.java

```diff
@@ -165,8 +159,18 @@ public static BaseLocale getInstance(String language, String script,
             language = convertOldISOCodes(language);
         }
 
-        Key key = new Key(language, script, region, variant, false);
-        return Cache.CACHE.get(key);
+        // Obtain the "normalized" BaseLocale, using un-normalized
+        // BaseLocale as the key. The returned "normalized" instance
+        // can subsequently be used by the Locale instance which
+        // guarantees the locale components are properly cased/interned.
+        synchronized (BaseLocale.class) {
```

Comment:
The simplification is good but I wonder if this coarse locking is going to be a problem, do we need to use some concurrent to avoid contention here?

Reply:
Replaced it with a ReentrantLock

Reply:
> Replaced it with a ReentrantLock

The concern is that this is a system-wide lock and so creates the potential for contention when many threads are bashing on Locale.of and other methods. Moving to use the JDK's ReferenceKeyMap with a CHM, or roll something similar, would avoid that but I can't helping feeling that the lookup is already a bit complicated. Right now, obtaining a Locale will create a non-normalized BaseLocale to use as a key, use the key to get (or compute) the normalized BaseLocale, then the normalized BaseLocale as a key for the Locale cache. Since it's early in JDK 22 it makes me wonder if we should step back and re-think this.

On the WHM, then it is a bit surprising that a Locales don't keep the non-normalized BaseLocale reachable. I assume this means that entries will be expunged from the CACHE if there is a reference processing between lookups. Also using weak refs makes it harder to reason about when the other part of the caching (in LocaleObjectCache) is used soft refs. I think this is another reason to think about a larger overhaul.

Reply:
One possibility would be to have the whole cache be soft-referenced, instead of weak-referencing the key.
Something like Reference<Map<BaseLocale, BaseLocale>> where Reference is a SoftReference and Map is a ConcurrentHashMap. 
But then you'd have to deal with a non-final field for the reference to the cache, and creating a new map and a new cache when the reference to the cache gets cleared.

Reply:
Jim's `java.lang.runtime.ReferenceKeyMap` would be a possible solution for it, although it is currently package private.